yarn-project/end-to-end/src/uniswap_trade_on_l1_from_l2.test.ts

```diff
@@ -271,4 +277,161 @@ describe('uniswap_trade_on_l1_from_l2', () => {
+    const daiL2BalanceBeforeSwap = await daiCrossChainHarness.getL2PublicBalanceOf(ownerAddress);
+
+    // 3. Owner gives uniswap approval to transfer funds on its behalf
+    const nonceForWETHTransferApproval = new Fr(2n);
+    const transferMessageHash = await hashPayload([
+      uniswapL2Contract.address.toField(),
+      wethCrossChainHarness.l2Token.address.toField(),
+      FunctionSelector.fromSignature('transfer_public((Field),(Field),Field,Field)').toField(),
+      ownerAddress.toField(),
+      uniswapL2Contract.address.toField(),
+      new Fr(wethAmountToBridge),
+      nonceForWETHTransferApproval,
+    ]);
+    await ownerWallet.setPublicAuth(transferMessageHash, true).send().wait();
+
+    // 4. Swap on L1 - sends L2 to L1 message to withdraw WETH to L1 and another message to swap assets.
+    const deadlineForDepositingSwappedDai = BigInt(2 ** 32 - 1); // max uint32 - 1
+    const [secretForDepositingSwappedDai, secretHashForDepositingSwappedDai] =
+      await daiCrossChainHarness.generateClaimSecret();
+
+    // 4.1 Owner approves user to swap on their behalf:
+    const nonceForSwap = new Fr(3n);
+    const swapMessageHash = await hashPayload([
+      user2Address.toField(),
+      uniswapL2Contract.address.toField(),
+      FunctionSelector.fromSignature(
+        'swap_public((Field),(Field),Field,(Field),Field,Field,Field,(Field),Field,Field,(Field),(Field),Field)',
+      ).toField(),
+      ownerAddress.toField(),
+      wethCrossChainHarness.l2Bridge.address.toField(),
+      new Fr(wethAmountToBridge),
+      daiCrossChainHarness.l2Bridge.address.toField(),
+      nonceForWETHTransferApproval,
+      new Fr(uniswapFeeTier),
+      new Fr(minimumOutputAmount),
+      ownerAddress.toField(),
+      secretHashForDepositingSwappedDai,
+      new Fr(deadlineForDepositingSwappedDai),
+      ownerEthAddress.toField(),
+      ownerEthAddress.toField(),
+      nonceForSwap,
+    ]);
+    await ownerWallet.setPublicAuth(swapMessageHash, true).send().wait();
+
+    // 4.2 Call swap_public from user2
+    const withdrawReceipt = await uniswapL2Contract
+      .withWallet(user2Wallet)
+      .methods.swap_public(
+        ownerAddress,
+        wethCrossChainHarness.l2Bridge.address,
+        wethAmountToBridge,
+        daiCrossChainHarness.l2Bridge.address,
+        nonceForWETHTransferApproval,
+        uniswapFeeTier,
+        minimumOutputAmount,
+        ownerAddress,
+        secretHashForDepositingSwappedDai,
+        deadlineForDepositingSwappedDai,
+        ownerEthAddress,
+        ownerEthAddress,
+        nonceForSwap,
+      )
+      .send()
+      .wait();
+    expect(withdrawReceipt.status).toBe(TxStatus.MINED);
+
+    // check weth balance of owner on L2 (we first bridged `wethAmountToBridge` into L2 and now withdrew it!)
+    await wethCrossChainHarness.expectPublicBalanceOnL2(ownerAddress, wethL2BalanceBeforeSwap - wethAmountToBridge);
+
+    // 5. Consume L2 to L1 message by calling uniswapPortal.swap()
```

Comment:
Maybe more saying if do something like
```suggestion
    // 5. Perform the swap on L1 with the `uniswapPortal.swap()` (consuming L2 to L1 messages)
```